discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,266 @@
+package zookeeper
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/joyent/containerpilot/discovery"
+	"github.com/joyent/containerpilot/utils"
+	"github.com/samuel/go-zookeeper/zk"
+)
+
+func init() {
+	discovery.RegisterBackend("zookeeper", ConfigHook)
+}
+
+// ZooKeeper wraps a ZooKeeper connection handler. It also stores the
+// prefix under which ContainerPilot nodes will be registered.
+type ZooKeeper struct {
+	Client    *zk.Conn
+	Prefix    string
+	eventChan chan zk.Event
+}
+
+// ServiceNode is the serializable form of a ZooKeeper service record
+type ServiceNode struct {
+	ID      string   `json:"id"`
+	Name    string   `json:"name"`
+	Address string   `json:"address"`
+	Port    int      `json:"port"`
+	Tags    []string `json:"tags"`
+}
+
+type zookeeperRawConfig struct {
+	Endpoints interface{} `mapstructure:"endpoints"`
+	Prefix    string      `mapstructure:"prefix"`
+}
+
+func (conn ZooKeeper) ttlHandler(service *discovery.ServiceDefinition, key string) {
+	for {
+		select {
+		case ev := <-conn.eventChan:
+			_, _, _, err := conn.Client.GetW(ev.Path)
+			if err != nil {
+				log.Warning(err)
+				conn.Deregister(service)
+				return
+			}
+		case <-time.After(time.Duration(service.TTL) * time.Second):
+			conn.Deregister(service)
+			return
+		}
+	}
+}
+
+func (conn ZooKeeper) eventChanCallBack(ev zk.Event) {
+	if ev.Type == zk.EventNodeDataChanged {
+		go func() {
+			conn.eventChan <- ev
+		}()
+	}
+}
+
+// ConfigHook is the hook to register with the ZooKeeper backend
+func ConfigHook(raw interface{}) (discovery.ServiceBackend, error) {
+	return NewZooKeeperConfig(raw)
+}
+
+func connection(addresses []string, cb zk.EventCallback) (*zk.Conn, error) {
+	c, _, err := zk.Connect(addresses, time.Second, zk.WithEventCallback(cb))
+	return c, err
+}
+
+// NewZooKeeperConfig creates a new service discovery backend for zookeeper
+func NewZooKeeperConfig(raw interface{}) (*ZooKeeper, error) {
+	zookeeper := &ZooKeeper{
+		Prefix:    "/containerpilot",
+		eventChan: make(chan zk.Event),
+	}
+	config := &struct {
+		Address string `mapstructure:"address"`
+	}{}
+	if err := utils.DecodeRaw(raw, &config); err != nil {
+		return nil, err
+	}
+	conn, err := connection([]string{config.Address}, zookeeper.eventChanCallBack)
+	if err != nil {
+		return nil, err
+	}
+	zookeeper.Client = conn
+	return zookeeper, nil
+}
+
+// Deregister removes this instance from the registry
+func (conn *ZooKeeper) Deregister(service *discovery.ServiceDefinition) {
+	key := conn.getNodeKey(service)
+	exists, _, err := conn.Client.Exists(key)
```

Comment:
The intent was to avoid excessive noise during deregistration, when we try to delete an already deregistered node. 

> Il giorno 12 set 2016, alle ore 15:43, Tim Gross <notifications@github.com> ha scritto:
> 
> In discovery/zookeeper/zookeeper.go:
> 
> > +	}{}
> > +	if err := utils.DecodeRaw(raw, &config); err != nil {
> > +		return nil, err
> > +	}
> > +	conn, err := connection([]string{config.Address}, zookeeper.eventChanCallBack)
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +	zookeeper.Client = conn
> > +	return zookeeper, nil
> > +}
> > +
> > +// Deregister removes this instance from the registry
> > +func (conn *ZooKeeper) Deregister(service *discovery.ServiceDefinition) {
> > +	key := conn.getNodeKey(service)
> > +	exists, _, err := conn.Client.Exists(key)
> Why check Client.Exists here (and throwing away the channel) and making an extra ZK request instead of just doing Client.Delete and then handling the error? You're throwing away the error in conn.Client.Delete too which means if it fails the check for Exists didn't do you any good anyways.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>